### ISSUE TITLE
Override default language packages in nixpkgs

### ISSUE THREAD

**Issue by tennox** [cachix/devenv#637]
Thanks for this great project! :heart: 
It's what I dreamt numtide/devshell could be & more :heart_eyes: 

I'm looking for a devenv-native, streamlined way to change nodejs version and also have the package managers use that version, for which the best solution I found (see [discussion in discourse](https://discourse.nixos.org/t/how-to-use-pnpm-with-recent-nodejs/21867/4?u=tennox)) was to use an overlay:

```nix 
{
  overlays = [
    (final: prev: {
      nodejs = cfg.package;
    })
  ];
}
```
And then all added packages use that, too:
```nix
{
  packages = with pkgs; [
    yarn
    nodePackages.pnpm
  ];
}
```
→ `pnpm node --version && yarn node --version` is the selected package

For supporting this in e.g. `languages/javascript.nix` all we'd need is a way to add an overlay in the language module - is there a way?

A possible option (but less great & flexible):
```nix
{
  languages = {
    javascript = {
      package-manager = "pnpm";
    };
  };
}
```

And then in the `languages/javascript.nix` if `packageManager` is `yarn` add `(yarn.override { nodejs = pkgs.nodejs-16_x; })`

Coming from #475 
Vaguely related to #16

**Comment by shyim** [comment]
I really like the idea of having an `package-manager` option in the javascript module and on non `npm` we could use nodejs slim to remove npm.

**Comment by soundofspace** [comment]
@tennox were you able to get these overlays working in devenv? I'm having a similar problem were even some package json dependencies fail because yarn is using the wrong node version. I'm currently still to novice at nix sadly to get it working myself.

**Comment by soundofspace** [comment]
Thanks will give that a go. Does this also work without flakes?
Just like you I came here since this is an amazing project (first tried devbox which is also an amazing project but sadly too basic for our needs). But I'm also worry about the extra complexity this introduces everywhere, as long as it works not really that big of an easy, but things like this make it tougher to sell.

**Comment by tennox** [comment]
> Thanks will give that a go. Does this also work without flakes?

I don't see a way without integrating this into devenv (which I meant to initiate with this issue, maybe I'll craft a PR 🧐)

> I'm also [worried] about the extra complexity this introduces everywhere, as long as it works not really that big of a deal, but things like this make it tougher to sell.

Exactly.

If it helps you, I can say that I have used devenv for many projects and almost never had to fiddle with/update the flake.nix in any 'difficult way' (only e.g. change nixpkgs channel/branch when a new stable nixos release came out)
Exceptions are nodejs version (see this issue) - and of course when building software via nix. (but that's just out of scope for devenv)

**Comment by domenkozar** [comment]
It seems there are two requests here:

a) supporting other package managers than `npm`, which we support now as per examples in https://github.com/cachix/devenv/tree/main/examples

b) supporting local overlays which we already have an issue for https://github.com/cachix/devenv/issues/478

**Comment by soundofspace** [comment]
@domenkozar  Problem is not really other package managers, problem is that package managers are using different nodejs version. Overlays are a way to fix this, but ideally it doesn't happen in the first place.
```  languages.javascript = {
    package = pkgs.nodejs_21;
    enable = true;
    # corepack.enable = true;
    yarn = {
      enable = true;
      install.enable = true;
    };
  };
  ```
  
  ```sh
 node --version
v21.7.2
yarn node --version
v20.11.1
```

**Comment by soundofspace** [comment]
This results in problems with packages such as better-sqlite3 when building dependencies:
```sh
yarn node -e 'console.log(process.config)'
-> node_module_version: 115,
node -e 'console.log(process.config)'
-> node_module_version: 120,
```
Error:
```
node_modules/better-sqlite3/build/Release/better_sqlite3.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 115. This version of Node.js requires
NODE_MODULE_VERSION 120. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`). {
  ```

**Comment by domenkozar** [comment]
Then we should always override that they use the canonical `package` of nodejs

**Comment by soundofspace** [comment]
Yes indeed. Seeing [this comment](https://github.com/cachix/devenv/issues/478#issuecomment-1663735284), would it be possible to do something similar for this snippet?
```
{
  overlays = [
    (final: prev: {
      nodejs = pkgs.nodejs_21;
    })
  ];
}
```

**Comment by bokhi** [comment]
recently started to work for a project that already had a `devcontainer` with specific versions of node, pnpm and deno. Here's what I did to get the exact same versions:

devenv.yaml
```yaml
# yaml-language-server: $schema=https://devenv.sh/devenv.schema.json
inputs:
  nixpkgs:
    url: github:cachix/devenv-nixpkgs/rolling
  # https://www.nixhub.io/packages/nodejs
  nodejs-18_18_0:
    url: github:NixOS/nixpkgs/517501bcf14ae6ec47efd6a17dda0ca8e6d866f9
  # https://www.nixhub.io/packages/nodePackages.pnpm
  pnpm-8_6_6:
    url: github:NixOS/nixpkgs/12d2f0da28a47f907825939630b4a0e8122fabe3
  # https://www.nixhub.io/packages/deno
  deno-1_39_0:
    url: github:NixOS/nixpkgs/2b9c57d33e3d5be6262e124fc66e3a8bc650b93d
```

devenv.nix
```nix
{ pkgs, lib, config, inputs, ... }:

let
  node-pkgs = import inputs.nodejs-18_18_0 { system = pkgs.stdenv.system; };
  pnpm-pkgs = import inputs.pnpm-8_6_6
    { system = pkgs.stdenv.system;
      overlays = [(final: prev: {nodePackages = prev.nodePackages.override { nodejs = node-pkgs.nodejs; }; })];
    };
   deno-pkgs = import inputs.deno-1_39_0 { system = pkgs.stdenv.system; };
in {
  languages.javascript.enable = true;
  languages.javascript.package = node-pkgs.nodejs;
  languages.javascript.pnpm.enable = true;
  languages.javascript.pnpm.package = pnpm-pkgs.nodePackages.pnpm;
  languages.javascript.pnpm.install.enable = true;

  languages.deno.enable = true;
  languages.deno.package = deno-pkgs.deno;
}
```

the difficult bit was to make sure that pnpm was using the same node version (`pnpm node -v` vs `node -v), was able to get that with an overlay and override.

**Comment by joscha** [comment]
vaguely related to #1382
